components/script/dom/performancenavigationtiming.rs

```diff
@@ -0,0 +1,115 @@
+    performanceresourcetiming: PerformanceResourceTiming,
+    navigation_start: u64,
+    navigation_start_precise: u64,
+    document: Dom<Document>,
+    nav_type: NavigationType,
+    redirect_count: u32,
+}
+
+impl PerformanceNavigationTiming {
+    fn new_inherited(nav_start: u64,
+                     nav_start_precise: u64,
+                     document: &Document)
+                         -> PerformanceNavigationTiming {
+        PerformanceNavigationTiming {
+            performanceresourcetiming: PerformanceResourceTiming::new_inherited(
+                document.url(),
+                DOMString::from("navigation"),
+                None,
+                nav_start_precise as f64),
+            navigation_start: nav_start,
+            navigation_start_precise: nav_start_precise,
+            document: Dom::from_ref(document),
+            nav_type: NavigationType::Navigate,
+            redirect_count: 0,
+        }
+    }
+
+    pub fn set_type(&mut self, nav_type: NavigationType) {
+        self.nav_type = nav_type
+    }
+}
+
+// https://w3c.github.io/navigation-timing/
+impl PerformanceNavigationTimingMethods for PerformanceNavigationTiming {
+    // https://w3c.github.io/navigation-timing/
+    fn UnloadEventStart(&self) -> DOMHighResTimeStamp {
+        Finite::wrap(self.document.get_unload_event_start() as f64)
+    }
+
+    // https://w3c.github.io/navigation-timing/
+    fn UnloadEventEnd(&self) -> DOMHighResTimeStamp {
+        Finite::wrap(self.document.get_unload_event_end() as f64)
+    }
+
+    // https://w3c.github.io/navigation-timing/
+    fn DomInteractive(&self) -> DOMHighResTimeStamp {
+        Finite::wrap(self.document.get_dom_interactive() as f64)
+    }
+
+    // https://w3c.github.io/navigation-timing/
+    fn DomContentLoadedEventStart(&self) -> DOMHighResTimeStamp {
+        Finite::wrap(self.document.get_dom_content_loaded_event_start() as f64)
+    }
+
+    // https://w3c.github.io/navigation-timing/
+    fn DomContentLoadedEventEnd(&self) -> DOMHighResTimeStamp {
+        Finite::wrap(self.document.get_dom_content_loaded_event_end() as f64)
+    }
+
+    // https://w3c.github.io/navigation-timing/
+    fn DomComplete(&self) -> DOMHighResTimeStamp {
+        Finite::wrap(self.document.get_dom_complete() as f64)
+    }
+
+    // https://w3c.github.io/navigation-timing/
+    fn LoadEventStart(&self) -> DOMHighResTimeStamp {
+        Finite::wrap(self.document.get_load_event_start() as f64)
+    }
+
+    // https://w3c.github.io/navigation-timing/
+    fn LoadEventEnd(&self) -> DOMHighResTimeStamp {
+        Finite::wrap(self.document.get_load_event_end() as f64)
+    }
+
+    // check-tidy: no specs after this line
+    // Servo-only timing for when top-level content (not iframes) is complete
+    fn TopLevelDomComplete(&self) -> DOMHighResTimeStamp {
+        Finite::wrap(self.document.get_top_level_dom_complete() as f64)
+    }
+
+    // TODO type
+    fn Type(&self) -> NavigationType {
```

Comment:
Let's move the APIs that are specced before the non-specced TopLevelDomComplete.